[FLINK-32161][test] Migrate and remove some legacy ExternalResource

reswqa opened the pull request:
## What is the purpose of the change

*Migrate and remove some legacy ExternalResource.*


## Brief change log

  - *Auto injects TestLoggerExtension for flink-tests.*
  - *Migrate and remove some legacy ExternalResource*

## Verifying this change


*This change is already covered by existing tests.*


## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

reswqa commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/ZooKeeperDefaultDispatcherRunnerTest.java:
```diff
@@ -47,58 +48,59 @@
 
 /** Tests for the interaction between the {@link DefaultDispatcherRunner} and ZooKeeper. */
-public class ZooKeeperDefaultDispatcherRunnerTest extends TestLogger {
+class ZooKeeperDefaultDispatcherRunnerTest {
```
`META-INF/services/org.junit.jupiter.api.extension.Extension` already exists in `flink-runtime` module. But exclude  it from the test-jar may need to be added.

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStoreTest.java:
```diff
@@ -101,13 +102,11 @@ public void testRecoverFailsIfDownloadFails() {
-                                        checkpointsInZooKeeper, zooKeeperCheckpointStoreUtil));
-        assertThat(exception, FlinkMatchers.containsCause(ExpectedTestException.class));
+                                        checkpointsInZooKeeper, zooKeeperCheckpointStoreUtil))
+                .satisfies(exp -> FlinkAssertions.anyCauseMatches(ExpectedTestException.class));
```
```suggestion
                .satisfies(FlinkAssertions.anyCauseMatches(ExpectedTestException.class));
```
lambda is probably useless here.
I tried replace `ExpectedTestException` with `NumberFormatException` and test still passes...
After removal of lambda it works better

reswqa replied:
Good catch 👍

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStoreTest.java:
```diff
@@ -232,12 +232,13 @@ public long getStateSize() {
+                HighAvailabilityOptions.HA_ZOOKEEPER_QUORUM,
+                zooKeeperExtensionWrapper.getCustomExtension().getConnectString());
 
         final CuratorFrameworkWithUnhandledErrorListener curatorFrameworkWrapper =
                 ZooKeeperUtils.startCuratorFramework(configuration, NoOpFatalErrorHandler.INSTANCE);
```
it seems here `ZooKeeperUtils.startCuratorFramework` opens resources and doesn't close it.. May be put it in `try` resources block

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/DefaultDispatcherRunnerITCase.java:
```diff
@@ -121,18 +120,20 @@ public void setup() {
     @Test
-    public void leaderChange_afterJobSubmission_recoversSubmittedJob() throws Exception {
+    void leaderChange_afterJobSubmission_recoversSubmittedJob() throws Exception {
         try (final DispatcherRunner dispatcherRunner = createDispatcherRunner()) {
```
just wonder whether we need `dispatcherRunner` here? It seems it is not used anywhere

reswqa replied:
We does need this `dispatcherRunner` as firstDispatcherGateway depend on and connect to it.

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/DefaultDispatcherRunnerITCase.java:
```diff
@@ -173,8 +175,7 @@ private DispatcherGateway electLeaderAndRetrieveGateway(UUID firstLeaderSessionI
+    void leaderChange_withBlockingJobManagerTermination_doesNotAffectNewLeader() throws Exception {
         final TestingJobMasterServiceLeadershipRunnerFactory jobManagerRunnerFactory =
                 new TestingJobMasterServiceLeadershipRunnerFactory(1);
         final TestingCleanupRunnerFactory cleanupRunnerFactory = new TestingCleanupRunnerFactory();
```
Here below similar question about `dispatcherRunner` whether we need it or not...
And at the same time there is 
```java
final TestingJobManagerRunner testingJobManagerRunner =
                    jobManagerRunnerFactory.takeCreatedJobManagerRunner();
```
where some resources are opened without closing... may be put it in `try` resource

reswqa replied:
> Here below similar question about dispatcherRunner whether we need it or not... (in case yes... may just add a comment)

It is also necessary for this test as `dispatcherRunner` will start `LeaderElectionService` and play the role of `LeaderContender`.

> where some resources are opened without closing... may be put it in try resource

Done.

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/ZooKeeperDefaultDispatcherRunnerTest.java:
```diff
@@ -144,8 +147,9 @@ public void teardown() throws Exception {
+                testingRpcServiceExtensionWrapper.getCustomExtension().getTestingRpcService();
         final TestingLeaderElectionService dispatcherLeaderElectionService =
                 new TestingLeaderElectionService();
 
```
should we also put `ZooKeeperUtils.startCuratorFramework` which is one line below to `try` resource?

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/ZooKeeperDefaultDispatcherRunnerTest.java:
```diff
@@ -232,7 +236,7 @@ public JobResultStore createJobResultStore() {
 
         // check resource clean up
-        assertThat(clusterHaStorageDir.listFiles(), is(emptyArray()));
+        assertThat(clusterHaStorageDir.listFiles()).isEmpty();
```
```suggestion
        assertThat(clusterHaStorageDir).isEmptyDirectory();
```

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/FileUploadHandlerITCase.java:
```diff
@@ -56,26 +56,26 @@
  * and/or json are properly handled.
  */
-public class FileUploadHandlerITCase extends TestLogger {
+class FileUploadHandlerITCase {
```
There are several tests in this class with `public` modifier which could be hardened

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/MultipartUploadExtension.java:
```diff
@@ -63,18 +64,18 @@
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 import static org.junit.Assert.assertArrayEquals;
 import static org.junit.Assert.assertEquals;
```
i guess it could be replaced with assertj
WDYT?

reswqa replied:
Yes, good catch! 👍

reswqa commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ZKCheckpointIDCounterMultiServersTest.java:
```diff
@@ -48,11 +49,12 @@ public final class ZKCheckpointIDCounterMultiServersTest extends TestLogger {
-                HighAvailabilityOptions.HA_ZOOKEEPER_QUORUM, zooKeeperResource.getConnectString());
+                HighAvailabilityOptions.HA_ZOOKEEPER_QUORUM,
+                zookeeperExtensionWrapper.getCustomExtension().getConnectString());
         final CuratorFrameworkWithUnhandledErrorListener curatorFrameworkWrapper =
                 ZooKeeperUtils.startCuratorFramework(configuration, NoOpFatalErrorHandler.INSTANCE);
 
```
Yes, this should work. I just migrated the testing and didn't do too much code refactoring.

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/InternalPriorityQueueTestBase.java:
```diff
@@ -261,33 +260,29 @@ public void testBulkAddRestoredElements() throws Exception {
+                    assertThat(elementSet.remove(iterator.next())).isTrue();
                 } else {
-                    Assert.assertTrue(elementSet.contains(iterator.next()));
+                    assertThat(elementSet.contains(iterator.next())).isTrue();
```
```suggestion
                    assertThat(elementSet).contains(iterator.next());
```

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/InternalPriorityQueueTestBase.java:
```diff
@@ -261,33 +260,29 @@ public void testBulkAddRestoredElements() throws Exception {
         }
         if (testSetSemanticsAgainstDuplicateElements()) {
-            Assert.assertTrue(elementSet.isEmpty());
+            assertThat(elementSet.isEmpty()).isTrue();
```
```suggestion
            assertThat(elementSet).isEmpty();
```

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/InternalPriorityQueueTestBase.java:
```diff
@@ -296,14 +291,14 @@ public void testIterator() throws Exception {
+                assertThat(checkSet.remove(iterator.next())).isTrue();
             }
-            Assert.assertTrue(checkSet.isEmpty());
+            assertThat(checkSet.isEmpty()).isTrue();
```
```suggestion
            assertThat(checkSet).isEmpty();
```

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/TaskExecutorLocalStateStoresManagerTest.java:
```diff
@@ -173,15 +180,16 @@ public void testLocalStateNoCreateDirWhenDisabledLocalRecovery() throws Exceptio
-                        .getLocalStateDirectoryProvider()
-                        .orElse(null));
+        assertThat(taskLocalStateStore.getLocalRecoveryConfig().isLocalRecoveryEnabled()).isFalse();
+        assertThat(
+                        taskLocalStateStore
+                                .getLocalRecoveryConfig()
+                                .getLocalStateDirectoryProvider()
+                                .orElse(null))
+                .isNull();
```
```suggestion
                assertThat(
                taskLocalStateStore
                        .getLocalRecoveryConfig()
                        .getLocalStateDirectoryProvider()).isNotPresent();
```

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/TaskExecutorLocalStateStoresManagerTest.java:
```diff
@@ -173,15 +180,16 @@ public void testLocalStateNoCreateDirWhenDisabledLocalRecovery() throws Exceptio
 
         for (File recoveryDir : rootDirs) {
-            Assert.assertEquals(0, recoveryDir.listFiles().length);
+            assertThat(recoveryDir.listFiles()).isEmpty();
```
```suggestion
            assertThat(recoveryDir).isEmptyDirectory();
```

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/TaskExecutorLocalStateStoresManagerTest.java:
```diff
@@ -53,25 +53,26 @@
 
 /** Test for {@link TaskExecutorLocalStateStoresManager}. */
-public class TaskExecutorLocalStateStoresManagerTest extends TestLogger {
+class TaskExecutorLocalStateStoresManagerTest {
```
In this class there are tests with public modifiers which could be hardened

reswqa replied:
I somehow missing this. 😞

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/heap/HeapPriorityQueueTest.java:
```diff
@@ -61,25 +62,25 @@ public void testToArray() {
-            Assert.assertEquals(timerPriorityQueue.size(), count);
-            Assert.assertTrue(checkSet.isEmpty());
+            assertThat(timerPriorityQueue.size()).isEqualTo(count);
+            assertThat(checkSet.isEmpty()).isTrue();
```
```suggestion
            assertThat(checkSet).isEmpty();
```

snuyanzin commented on flink-tests/src/test/java/org/apache/flink/test/recovery/ProcessFailureCancelingITCase.java:
```diff
@@ -214,12 +217,14 @@ public Long map(Long value) throws Exception {
-            assertTrue(error instanceof ProgramInvocationException);
+            assertThat(programException.get())
+                    .withFailMessage("The program did not fail properly")
+                    .isNotNull()
```
```suggestion
```
no need for `isnotNull` since it's covered by `isInstanceOf`

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/heap/HeapPriorityQueueTest.java:
```diff
@@ -61,25 +62,25 @@ public void testToArray() {
             HeapPriorityQueue<TestElement> timerPriorityQueue = newPriorityQueue(1);
 
-            Assert.assertEquals(testArray.length, timerPriorityQueue.toArray(testArray).length);
+            assertThat(testArray).hasSameSizeAs(timerPriorityQueue);
```
```suggestion
            assertThat(testArray).hasSameSizeAs(timerPriorityQueue.toArray(testArray));
```
Seems it is the reason of CI failure

reswqa commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/NettyLeakDetectionExtension.java:
```diff
@@ -85,7 +87,7 @@ protected synchronized void after() {
         public <T> ResourceLeakDetector<T> newResourceLeakDetector(
                 Class<T> resource, int samplingInterval, long maxActive) {
-            return new FailingResourceLeakDetector<T>(resource, samplingInterval, maxActive);
+            return new FailingResourceLeakDetector<>(resource, samplingInterval, maxActive);
```
Good suggestion.

reswqa commented:
@snuyanzin Thank you for reviewing this so carefully! I have to admit that I was not in a good state when doing this, so many assertion rewrites made me feel a bit dizzy 😵. I am very sorry for some careless and foolish things like forgetting to remove all public modifier. 🙇. 

I have addressed your comments in some fix-up commits and put the hotfix commit of excluding `META-INF/services/org.junit.jupiter.api.extension.Extension` from `flink-runtime` test-jar in the first commit.

reswqa commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/ZooKeeperDefaultDispatcherRunnerTest.java:
```diff
@@ -153,15 +154,17 @@ void testResourceCleanupUnderLeadershipChange() throws Exception {
-                        .setJobMasterLeaderRetrieverFunction(
-                                jobId -> ZooKeeperUtils.createLeaderRetrievalService(client))
-                        .build()) {
+        try (final CuratorFrameworkWithUnhandledErrorListener curatorFrameworkWrapper =
+                        ZooKeeperUtils.startCuratorFramework(configuration, fatalErrorHandler);
```
The following way is not used here as it will pop up a warning: `CuratorFrameworkWithUnhandledErrorListener used without 'try'-with-resources statement`.

```
try (CuratorFramework curatorFramework =
                        ZooKeeperUtils.startCuratorFramework(configuration, fatalErrorHandler)
                                .asCuratorFramework();
```

snuyanzin commented:
No worries, everything is ok
thanks a lot for addressing my comments.

It looks ok from my side

P.S. about assertions: there is Concise AssertJ Optimizing Nitpicker plugin [1] which could help  with hints. It has some bugs (so it is not perfect) however it helps a lot

[1] https://plugins.jetbrains.com/plugin/12195-concise-assertj-optimizing-nitpicker-cajon-

reswqa commented:
Thanks for the feedback. That's really interesting, I'll try installing this plugin 😆

reswqa commented:
Rebased all fix-up commits, let's waiting for AZP green.
